### ISSUE TITLE
Livecheck: extract_plist allow specific filepath to check

### ISSUE THREAD

**Issue by bevanjkay** [Homebrew/brew#11619]
### Provide a detailed description of the proposed feature

Allow the ability to specify the path of an application/package to check the info.plist

### What is the motivation for the feature?

In the case of the cask for Amazon Music, the version is stored in the .plist file in the installer application. As this application is not added as an artifact, it is not used in the livecheck.
The app could be provided in the livecheck block.

https://github.com/Homebrew/homebrew-cask/pull/107834

### How will the feature be relevant to at least 90% of Homebrew users?

Will allow coverage of some additional livecheck cases

### What alternatives to the feature have been considered?

Considering alternative ways to run livecheck for this and other casks, but no solutions found as yet.

**Comment by MikeMcQuaid** [comment]
CC @maxim-belkin @samford @nandahkrishna for livecheck experts' thoughts.

**Comment by samford** [comment]
I'll respond to this later today after I've finished collecting data on it. I didn't get to review this strategy when it was originally added (I missed it and it was merged without any review and none was requested) and I'm not convinced the approach it uses is appropriate for livecheck (it's incredibly data-intensive and also quite slow). I'll elaborate when I respond later, though.

**Comment by maxim-belkin** [comment]
`amazon-music` doesn't have a livecheck block atm. When I plug `extract_plist` strategy, `brew livecheck amazon-music` fails. The failure happens on line 

```ruby
versions = unversioned_cask_checker.all_versions.transform_values { |v| Item.new(bundle_version: v) }
```
in `extract_plist.rb`, indicating that it's the `all_versions` from `unversioned_cask_checker.rb` that is failing and not the strategy itself. So, I think the right fix is to make `all_versions` method in `unversioned_cask_checker.rb` work for `amazon-music`

**Comment by bevanjkay** [comment]
@maxim-belkin yes I did make it about that far in my digging. In the case of Amazon Music the .dmg container has just the one .app (Amazon Music Installer.app) which does contain an info.plist with the correct version. 

If I add an app stanza to the cask including this app (`app 'Amazon Music Installer.app'`) then livecheck finds it. I don't believe it would be good to look in any .app present, and this is where providing an app to check could be valuable. 


@samford Your thoughts are certainly accurate. Extract Plist can be quite data intensive as it fetches the file during the livecheck. In the case of Amazon Music that results in a decently sized download (~80MB). When we do use the strategy it is as an absolute last resort.

**Comment by maxim-belkin** [comment]
@samford, did you need more time to look into this?

> I don't believe it would be good to look in any .app present, and this is where providing an app to check could be valuable.

I agree.

**Comment by samford** [comment]
> @samford, did you need more time to look into this?

Ah, no, I had finished collecting information but became busy and forgot to come back to this. My general concerns with the `ExtractPlist` strategy don't affect this issue directly, so feel free to move this forward in #11667. I don't have much knowledge of how the cask side of Homebrew works, so I'll leave review of that PR up to homebrew/cask maintainers (as this isn't strictly a livecheck issue).

---

As an aside, my main issue with `ExtractPlist` is that the 47 casks that use it can download ~3 GB of data and take 15 minutes in total to run these checks and this will become even worse if additional casks use it over time. For comparison, the ~5300 working checks in homebrew/core only download ~104 MB of data and it takes about an hour to run. The ~3500 working checks in homebrew/cask that don't use `ExtractPlist` only download ~60 MB of data and take ~45 minutes.

Before digging into how `ExtractPlist` works, I had simply noticed that it was very slow and didn't know that it downloaded the same large files used to install the cask. I don't imagine that many livecheck users understand that running `brew livecheck --tap=homebrew/cask` will opt them into downloading up to 3 GB of data. The only saving grace is that these files are cached, so follow-up runs won't use as much data.

I'm glad this strategy is only used as an absolute last resort but I think we'll still need to add something like `--only`/`--except` to livecheck in the future, to filter which formulae/casks are checked in a given run based on the strategy they use (e.g., `brew livecheck --tap=homebrew/cask --except=extract_plist`). Being able to implement this depends on some refactoring I have planned in support of another livecheck feature, so I'll work on this sometime after that's done. In the interim time, I'll simply have to resist the urge to periodically clean up my Homebrew cache.

**Comment by bevanjkay** [comment]
Thanks @samford 

#11667 adds functionality in a related but different area. I'll have a look at a PR for this issue also when I have time.

**Comment by samford** [comment]
> Given that extract_plist is generally data heavy, would it be worth discussing having it not run by default and be behind a flag (we can enable in CI)?

That was my first thought (e.g., using a generic flag like `--intensive-strategies` to opt-in to `ExtractPlist`) but I was concerned that it would be a more complicated mental model to grasp.

For example, if someone runs `--tap=homebrew/cask`, it's reasonable to expect that livecheck will check all the casks in the tap. In the case of homebrew/cask, livecheck would skip `ExtractPlist` checks unless the extra flag is used. This behavior wouldn't be obvious before running livecheck, so some folks may be [gently] bitten by it before they're told how to avoid it (e.g., in output).

Additionally, it could conflict with an `--except-strategies` option in the future. `--except-strategies` would imply that all strategies except the supplied one(s) (e.g., `--except-strategies=npm,sourceforge`) would be used and that may not be true for `ExtractPlist`. It's not immediately clear to users whether `--except-strategies` would also skip `ExtractPlist` by default, so it makes livecheck's behavior harder to understand.

On the other hand, it may be that it's primarily maintainers who are interested in running *all* the checks (even the intensive ones) and your average user may be better served by not running `ExtractPlist` by default. If we provided some default output making it clear that `ExtractPlist` is skipped unless the flag is used, that would inform users about what's happening and how to avoid it.

Past that, this flag is also kind of similar to `--online` for `brew tests` (where tests that make certain network requests are skipped unless the flag is used), so there may be precedent for this kind of behavior.

---

In general, making `ExtractPlist` opt-in could be better for most users, who aren't likely to be aware of how data intensive the strategy is. It's more likely that knowledgeable livecheck users (e.g., maintainers) would know to use an `--intensive-strategies` flag to opt-in to using `ExtractPlist` when needed, rather than expecting your average user to know enough to want to opt-out using `--except-strategies` in the future. Using the flag on CI would essentially maintain the existing behavior in homebrew/cask.

However, I'm not fond of how it would make livecheck's behavior less predictable and harder to understand. I'm still on the fence about it but, since `ExtractPlist` has to be specified in a `livecheck` block, it's relatively easy to implement this idea and we could discuss the benefits/detriments in a PR. I'm working on a higher-priority feature right now but I can throw it together sometime afterward.

**Comment by maxim-belkin** [comment]
> On the other hand, it may be that it's primarily maintainers who are interested in running _all_ the checks

The `livecheck` command lives in `dev-cmd` directory, so I think this assumption is correct.
